src/SignalR/clients/cpp/src/signalrclient/connection_impl.cpp

```diff
@@ -84,94 +84,139 @@ namespace signalr
-                if (!connection)
+                return pplx::task_from_exception<negotiation_response>(_XPLATSTR("connection no longer exists"));
+            }
+            return request_sender::negotiate(*connection->m_web_request_factory, url,
+                connection->m_query_string, connection->m_signalr_client_config);
+        }, m_disconnect_cts.get_token())
+            .then([weak_connection, start_tce, redirect_count, url](negotiation_response negotiation_response)
+        {
+            auto connection = weak_connection.lock();
+            if (!connection)
+            {
+                return pplx::task_from_exception<void>(_XPLATSTR("connection no longer exists"));
+            }
+
+            if (!negotiation_response.error.empty())
+            {
+                return pplx::task_from_exception<void>(signalr_exception(negotiation_response.error));
+            }
+
+            if (!negotiation_response.url.empty())
+            {
+                if (!negotiation_response.accessToken.empty())
                 {
-                    return pplx::task_from_exception<negotiation_response>(_XPLATSTR("connection no longer exists"));
+                    auto headers = connection->m_signalr_client_config.get_http_headers();
+                    headers.add(_XPLATSTR("Authorization"), _XPLATSTR("Bearer ") + negotiation_response.accessToken);
+                    connection->m_signalr_client_config.set_http_headers(headers);
                 }
-                return request_sender::negotiate(*connection->m_web_request_factory, connection->m_base_url,
-                    connection->m_query_string, connection->m_signalr_client_config);
-            }, m_disconnect_cts.get_token())
-            .then([weak_connection](negotiation_response negotiation_response)
+                return connection->start_negotiate(negotiation_response.url, redirect_count + 1);
+            }
+
+            connection->m_connection_id = std::move(negotiation_response.connectionId);
+
+            // TODO: fallback logic
+
+            bool foundWebsockets = false;
+            for (auto availableTransport : negotiation_response.availableTransports)
             {
-                auto connection = weak_connection.lock();
-                if (!connection)
+                if (availableTransport.transport == _XPLATSTR("WebSockets"))
                 {
-                    return pplx::task_from_exception<void>(_XPLATSTR("connection no longer exists"));
+                    foundWebsockets = true;
+                    break;
                 }
-                connection->m_connection_id = std::move(negotiation_response.connection_id);
+            }
 
-                // TODO: check available transports
+            if (!foundWebsockets)
+            {
+                return pplx::task_from_exception<void>(signalr_exception(_XPLATSTR("WebSockets is the only supported transport currently")));
```

Comment:
Nit: "WebSockets is the only supported transport currently" -> "The server does not support WebSockets which is currently the only transport supported by this client."